nixos/modules/services/web-servers/authelia.nix

```diff
@@ -0,0 +1,181 @@
+{ lib
+, pkgs
+, config
+, ...
+}:
+
+let
+  cfg = config.services.authelia;
+
+  format = pkgs.formats.yaml { };
+  configFile = format.generate "config.yml" cfg.settings;
+
+  autheliaOpts = with lib; { name, ... }: {
+    options = {
+      enable = mkEnableOption "Authelia instance";
+
+      name = mkOption {
+        type = types.str;
+        default = name;
+      };
+
+      package = mkOption {
+        default = pkgs.authelia;
+        type = types.package;
+        defaultText = literalExpression "pkgs.authelia";
+      };
+
+      user = mkOption {
+        default = "authelia-${name}";
+        type = types.str;
+      };
+      group = mkOption {
+        default = "authelia-${name}";
+        type = types.str;
+      };
+
+      jwtSecretFile = mkOption {
+        type = types.path;
+        default = null;
+        description = ''
+          Path to your JWT secret used during identity verificaiton.
+        '';
+      };
+
+      oidcIssuerPrivateKeyFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = ''
+          Path to your private key file used to encrypt OIDC JWTs.
+        '';
+      };
+
+      oidcHmacSecretFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = ''
+          Path to your HMAC secret used to sign OIDC JWTs.
+        '';
+      };
+
+      sessionSecretFile = mkOption {
+        type = types.nullOr types.path;
+        default = null;
+        description = ''
+          Path to your session secret. Only used when redis is used as session storage.
+        '';
+      };
+
+      storageEncryptionKeyFile = mkOption {
+        type = types.path;
+        default = null;
+        description = ''
+          Path to your storage encryption key.
+        '';
+      };
+
+      settings = mkOption {
```

Comment:
`settings` option can be supplemented by an additional `settingsFile` option. It is useful, if you want to add client secrets, but don't want the to be visible in code.

https://github.com/LongerHV/nixos-configuration/blob/75f48d7d6853d355b8ad46cc1cda09436b97ed03/modules/nixos/authelia.nix#L25

Reply:
Well the client secrets won't be within nix but the secrets will still be in the config file which is still not really recommended. I did intend to add a settingsFile option though.

https://www.authelia.com/configuration/methods/secrets

I'll add that now and I think I'll add a sanity check

Reply:
Unfortunatelly client secrets must be defined in the configuration file, see this [issue](https://github.com/authelia/authelia/issues/3249).
I have most configuration in plain nix, but client specific stuff is encrypted with agenix and passed to authelia as a second config file.

Reply:
interesting. yeah at least you can provide authelia multiple config files (added a `settingsFiles` array)
just finishing up some bits